tensorboard/webapp/tbdev_upload/tbdev_upload_dialog_component.ts

```diff
@@ -12,20 +12,61 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
-import {Component} from '@angular/core';
+import {Component, OnInit, OnDestroy} from '@angular/core';
 import {MatDialogRef} from '@angular/material/dialog';
+import {Store, select} from '@ngrx/store';
+import {State, getEnvironment} from '../core/store';
+import {Subject} from 'rxjs';
+import {takeUntil} from 'rxjs/operators';
+
+/** @typehack */ import * as _typeHackRxjs from 'rxjs';
 
 @Component({
   selector: 'tbdev-upload-dialog',
   templateUrl: './tbdev_upload_dialog_component.ng.html',
   styleUrls: ['./tbdev_upload_dialog_component.css'],
 })
-export class TbdevUploadDialogComponent {
-  readonly commandText: string = 'tensorboard dev upload --logdir {logdir}';
+export class TbdevUploadDialogComponent implements OnInit, OnDestroy {
+  commandText: string = this.getCommandText('');
+  readonly environment$ = this.store.pipe(select(getEnvironment));
+
+  private ngUnsubscribe = new Subject();
+
+  constructor(
+    private dialogRef: MatDialogRef<TbdevUploadDialogComponent>,
+    private store: Store<State>
+  ) {}
+
+  ngOnInit() {
+    this.environment$
+      .pipe(takeUntil(this.ngUnsubscribe))
+      .subscribe((environment) => {
+        if (environment && environment.data_location) {
+          this.commandText = this.getCommandText(environment.data_location);
+        }
+      });
+  }
 
-  constructor(private dialogRef: MatDialogRef<TbdevUploadDialogComponent>) {}
+  ngOnDestroy() {
+    this.ngUnsubscribe.next();
+    this.ngUnsubscribe.complete();
+  }
 
   onClose(): void {
     this.dialogRef.close();
   }
+
+  private getCommandText(logdir?: string) {
+    if (!logdir) {
+      // Without logdir we print a literal '{logdir}' string that user will
+      // have to manually substitute.
+      return 'tensorboard dev upload --logdir {logdir}';
+    } else {
+      // With logdir we substitute the value into the command for the user.
+      // We assume that logdir is sufficiently long that we want to print it on
+      // the next line. If the logdir value is still too long then the CSS will
+      // render a scrollbar underneath.
+      return 'tensorboard dev upload --logdir \\\n    ' + logdir;
```

Comment:
As written, this isn’t escaped, and so the command can be wrong (if the
logdir contains spaces or some special characters) or even malicious:

![Screenshot of code injection with logdir named `--nope || ls /`][ss]

[ss]: https://user-images.githubusercontent.com/4317806/86337799-6f382e80-bc06-11ea-87c5-4c8cf24da2dc.png

Can we escape for POSIX `sh` here? (Probably wouldn’t be exactly right
on Windows, but that’s fine with me.)

Reply:
Thanks for taking a close look. Would it be sufficient to wrap the logdir value in quotes?

Reply:
I gave this a shot: Wrap the logdir value in (single) quotes and escape quote literals in the logdir.  I added a test for it, too.

Reply:
William, in what case does the logdir from `environment` contain such command? Does our tf-runs-selector handle the case properly?

Reply:
@bmd3k: Yep, that should work on POSIX shells; thanks!

@stephanwlee: I ran with `--logdir '--nope || ls /'` to get that
screenshot. The runs selector shows this path correctly and without
escaping (see bottom-left of screenshot), which is fine imho since it’s
not intended to be run as a command.